processor.go

```diff
@@ -136,19 +150,51 @@ func (p *ServiceProcessor) ProcessClientRequest(req *http.Request, path string,
 		arg.Elem().Set(reflect.ValueOf(msg).Elem())
 		ret := f.Call([]reflect.Value{arg})
 
+		// there will be an extra element in ret
 		ierr := ret[1].Interface()
 		if ierr != nil {
 			return nil, ierr.(error)
 		}
 		return ret[0].Interface(), nil
 	}()
 	if err != nil {
-		return nil, err
+		return nil, nil, err
 	}
+
+	if reply != nil && reflect.TypeOf(reply).Kind() == reflect.Chan {
+		outChan := make(chan []byte, 100)
+		c := reflect.ValueOf(reply)
+		case0 := reflect.SelectCase{Dir: reflect.SelectRecv, Chan: c}
+		go func() {
+			for {
+				chosen, v, ok := reflect.Select([]reflect.SelectCase{case0})
+				if !ok {
+					log.Info("the channel is closed")
+					return
+				}
+				if chosen == 0 {
+					buf, err = protobuf.Encode(v.Interface())
+					if err != nil {
+						log.Error(err)
+						// TODO close these?
+						close(outChan)
+						case0.Chan.Close()
+						return
+					}
+					outChan <- buf
+				}
+				// We don't add a way to close here, otherwise
```

Comment:
Yes, these issues are going to be the hardest. Starting streaming is easy, knowing when to stop is hard.

We need that an error on write down towards the ws client results in closing the channel, so that the service can stop sending.

And we'll need to call SetWriteDeadline on the ws, so that if the client forgets to read from us, it cannot hold open the server-side resources.